kernel_spark/src/main/scala/io/delta/ExpressionUtils.scala

```diff
@@ -0,0 +1,203 @@
+  private def convertKtoSExpr(kernelExpression: KernelExpression): Option[SparkExpression] = {
+    val result = kernelExpression match {
+      case expr: KernelAnd =>
+        for {
+          left <- convertKtoSPredicate(expr.getLeft)
+          right <- convertKtoSPredicate(expr.getRight)
+        } yield new SparkAnd(left, right)
+
+      case expr: KernelOr =>
+        for {
+          left <- convertKtoSPredicate(expr.getLeft)
+          right <- convertKtoSPredicate(expr.getRight)
+        } yield new SparkOr(left, right)
+
+      case expr: KernelPredicate if KernelPredicate.BINARY_OPERATORS.contains(expr.getName) =>
+        for {
+          left <- convertKtoSExpr(expr.getChildren.get(0))
+          right <- convertKtoSExpr(expr.getChildren.get(1))
+        } yield new SparkPredicate(expr.getName, Array(left, right))
+
+      case expr: KernelPredicate if KernelPredicate.UNARY_OPERATORS.contains(expr.getName) =>
+        for {
+          child <- convertKtoSExpr(expr.getChildren.get(0))
+        } yield new SparkPredicate(expr.getName, Array(child))
+
+      case expr: KernelColumn => Some(Expressions.column(expr.getNames.mkString(".")))
+
+      case literal: KernelLiteral =>
+        literal.getDataType match {
+          case _: kerneltypes.BooleanType =>
+            Some(Expressions.literal(literal.getValue.asInstanceOf[Boolean]))
+          case _: kerneltypes.IntegerType =>
+            Some(Expressions.literal(literal.getValue.asInstanceOf[Integer]))
+          case _: kerneltypes.LongType =>
+            Some(Expressions.literal(literal.getValue.asInstanceOf[Long]))
+          case _: kerneltypes.StringType =>
+            Some(Expressions.literal(literal.getValue.asInstanceOf[String]))
+          case _ =>
+            throw new UnsupportedOperationException(
+              s"Unsupported KernelLiteral type: ${literal.getDataType}")
+        }
+
+      case x =>
+        logger.warn(s"Unsupported KernelExpression: $x")
+        throw new UnsupportedOperationException(s"Unsupported KernelExpression: $x")
+    }
+
+    logger.info(s"convertKtoSExpr: input=$kernelExpression, result=$result")
+
+    result
+  }
+
+  //////////////////////
+  // Spark --> Kernel //
+  //////////////////////
+
+  // TODO: perhaps this should NOT return an Option?
+
+  def convertStoKPredicate(sparkPredicate: SparkPredicate): Option[KernelPredicate] = {
+    convertStoKExpr(sparkPredicate) match {
+      case Some(pred: KernelPredicate) => Some(pred)
+      case Some(expr) =>
+        throw new IllegalArgumentException(
+          s"Expected a KernelPredicate but got ${expr.getClass.getSimpleName}")
+      case _ => None
+    }
+  }
+
+  private def convertStoKExpr(sparkExpression: SparkExpression): Option[KernelExpression] = {
+    sparkExpression match {
+      case expr: SparkAnd =>
+        for {
+          left <- convertStoKPredicate(expr.left())
+          right <- convertStoKPredicate(expr.right())
+        } yield new KernelAnd(left, right)
```

Comment:
Is this an obscure way to map over a pair of options?
```suggestion
        convertStoKPredicate(expr.left())
          .zip(convertStoKPredicate(expr.right()))
          .map((left, right) => new KernelAnd(left, right))
```